pallets/bonds/src/lib.rs

```diff
@@ -2208,32 +2215,35 @@ decl_module! {
             Ok(())
         }
         
-        ///Adding funds to the insurer reserves
-        ///Needs amount transfer implemented
+        ///Adding to the balance of InsurerReserves and freezing Asset Id
         #[weight = 1000]
-        pub fn stake(origin, id: u32, deposit: u32) -> dispatch::DispatchResult {
-            let signer = ensure_signed(origin)?;
-            ensure!(InsurerReserves::<T>::contains_key(&signer, id.clone()), Error::<T>::ReserveNotFound);
+        pub fn stake(origin, id: u32, deposit: u32) -> DispatchResultWithPostInfo {
+            let signer = ensure_signed(origin.clone())?;
+            ensure!(InsurerReserves::<T>::contains_key(signer.clone(), id.clone()), Error::<T>::ReserveNotFound);
+            //Freezing this Asset Id before the change in reserves are made
+            pallet_assets::Module::<T>::freeze_asset(origin.clone(), id.clone())?;
             let reserve = InsurerReserves::<T>::take(signer.clone(), id.clone());
             let new_reserve: u128 = reserve.checked_add(deposit.into()).unwrap();
             InsurerReserves::<T>::insert(signer.clone(), id.clone(), new_reserve);
-            Ok(())
+            Ok(().into())
         }
         
-        ///Removing funds only if the reserve is at minimum required amount
-        ///Needs amount transfer implemented
+        ///Removing funds only if the reserve is at minimum required amount and thawing Asset Id
         #[weight = 1000]
-        pub fn unstake(origin, id: u32, withdrawal: u32) -> dispatch::DispatchResult {
-            let signer = ensure_signed(origin)?;
-            ensure!(InsurerReserves::<T>::contains_key(&signer, id.clone()), Error::<T>::ReserveNotFound);
+        pub fn unstake(origin, id: u32, withdrawal: u32) -> DispatchResultWithPostInfo {
+            let signer = ensure_signed(origin.clone())?;
+            ensure!(InsurerReserves::<T>::contains_key(signer.clone(), id.clone()), Error::<T>::ReserveNotFound);
+            //Retrieve the current minimum reserve required
             let settings_reserve: Vec<u8> = Settings::get("insuranceminreserve".as_bytes().to_vec()).unwrap();
             let reserve = json_get_value(settings_reserve.clone(), "reserve".as_bytes().to_vec());
             let reserve_min = vecu8_to_u32(reserve);
             let current_reserves = InsurerReserves::<T>::take(signer.clone(), id);
             ensure!(current_reserves >= reserve_min.into(), Error::<T>::BelowMinimumReserve);     
             let new_reserve: u128 = current_reserves.checked_sub(withdrawal.into()).unwrap();
             InsurerReserves::<T>::insert(signer.clone(), id.clone(), new_reserve);
-            Ok(())
+            //Unfreezing of this particular Asset id
+            pallet_assets::Module::<T>::thaw_asset(origin.clone(), id.clone())?;
```

Comment:
you must unlock the amount not the whole asset id

Reply:
`freeze_asset` and `thaw_asset` will not be considered after this.

Reply:
please think about the logic of  certain amount in reserve to cover for the risks of an insurance.